src/widgets/index/index.ts

```diff
@@ -25,6 +25,82 @@ import {
   mergeSearchParameters,
 } from '../../lib/utils';
 
+import { WidgetType, WidgetParams, Schema } from '../../../schema';
+
+function Pbf() {
+  this.buf = new Uint8Array(0);
+  this.pos = 0;
+  this.type = 0;
+  this.length = 0;
+}
+
+Pbf.prototype = {
+  finish() {
+    this.length = this.pos;
+    this.pos = 0;
+    return this.buf.subarray(0, this.length);
+  },
+
+  writeVarint(val) {
+    val = +val || 0;
+
+    if (val > 0xfffffff || val < 0) throw new Error('Unsupported');
+
+    // realloc
+    let length = this.length || 16;
+    while (length < this.pos + 4) length *= 2;
+    if (length !== this.length) {
+      const buf = new Uint8Array(length);
+      buf.set(this.buf);
+      this.buf = buf;
+      this.length = length;
+    }
+
+    this.buf[this.pos++] = (val & 0x7f) | (val > 0x7f ? 0x80 : 0);
+    if (val <= 0x7f) return;
+    this.buf[this.pos++] = ((val >>>= 7) & 0x7f) | (val > 0x7f ? 0x80 : 0);
+    if (val <= 0x7f) return;
+    this.buf[this.pos++] = ((val >>>= 7) & 0x7f) | (val > 0x7f ? 0x80 : 0);
+    if (val <= 0x7f) return;
+    this.buf[this.pos++] = (val >>> 7) & 0x7f;
```

Comment:
what does this code mean? It's quite dense

Reply:
Oh yeah, do not focus too much on it. It is a stripped down version of [pbf](https://github.com/mapbox/pbf) (a protobuf implementation) with only the methods that are useful to us to keep it as light as possible (since the original implementation is not tree-shakable).